fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -885,6 +954,56 @@ protected void validateSameMetadataOrThrow(FDBRecordStore store) {
         }
     }
 
+    CompletableFuture<Void> indexingStamp(@Nonnull ConcurrentHashMap<String, IndexBuildProto.IndexBuildIndexingStamp> oldStamps,
+                                          @Nullable OnlineIndexer.IndexingStampOperation op,
+                                          @Nullable String id,
+                                          @Nullable  Integer timeoutSeconds) {
+        return getRunner().runAsync(context -> openRecordStore(context).thenCompose(store ->
+            forEachTargetIndex(index -> store.loadIndexingTypeStampAsync(index)
+                    .thenApply(stamp -> indexingStamp(oldStamps, store, index, stamp, op, id, timeoutSeconds)))
+        ));
+    }
+
+    boolean indexingStamp(@Nonnull ConcurrentHashMap<String, IndexBuildProto.IndexBuildIndexingStamp> oldStamps,
+                          @Nonnull FDBRecordStore store,
+                          @Nonnull Index index,
+                          @Nullable IndexBuildProto.IndexBuildIndexingStamp stamp,
+                          @Nullable OnlineIndexer.IndexingStampOperation op,
+                          @Nullable String id,
+                          @Nullable  Integer timeoutSeconds
+    ) {
+        oldStamps.put(index.getName(), stamp != null ? stamp :
+                                       IndexBuildProto.IndexBuildIndexingStamp.newBuilder()
+                                               .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.INVALID)
+                                               .build());
+        if (op == null || stamp == null || op.equals(OnlineIndexer.IndexingStampOperation.QUERY)) {
+            return false;
+        }
+
+        IndexBuildProto.IndexBuildIndexingStamp.Builder builder;
+        try {
+            builder = IndexBuildProto.IndexBuildIndexingStamp.parseFrom(stamp.toByteArray()).toBuilder();
+        } catch (InvalidProtocolBufferException e) {
+            return false;
+        }
+
+        if (op == OnlineIndexer.IndexingStampOperation.BLOCK) {
+            builder.setBlock(true);
+            if (id != null) {
```

Comment:
🤷 
I really don't know. This is the first version - I tried to keep everything possible. But sure, we can enforce id to avoid having a legacy of `block=true, blockId=<undef>`. 
Note that the id may still be an empty string.